### PR TITLE
[mhlo] modify rng algrothim attr declare as enum 

### PR DESCRIPTION
**Pull request by lipracer**
[modify rng algrothim attr declare as enum](https://github.com/tensorflow/tensorflow/pull/55752/commits/6cfd66fc0d53f9e4a27aea1f8671bd65ce7394d6)

### CODE REVIEW

**Comment by jpienaar** [comment]
Thanks for the attribute enum change.

This rng op contrary to the other two doesn't have side effects/is stateless. It takes state and returns state but the op itself has no internal state or mutation given an input state it should produce the same output state so if one wants different values one has to thread the state around. Do you have an example where CSE is changing the result?

**Comment by lipracer** [comment]
> 
I will add custom builder function to fix compile error.

Consider the following IR if the RNG is really random, whether the result of the two random times is the same is what we expect? I am not particularly clear about the semantics of this Operation, if there are mistakes, thank you for correcting.

```
// RUN: mlir-hlo-opt %s -cse -allow-unregistered-dialect | FileCheck %s

func.func @rng_bit_generator(%arg0: tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>) {
  %0 = mhlo.constant dense<[10, 12]> : tensor<2xui64>
  %1, %2 = "mhlo.rng_bit_generator"(%0) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
  %3, %4 = "mhlo.rng_bit_generator"(%0) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
  func.return %1, %2, %3, %4 : tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>
}
```

**Comment by lipracer** [comment]
> > 
> 
> I will add custom builder function to fix compile error.
> 
> Consider the following IR if the RNG is really random, whether the result of the two random times is the same is what we expect? I am not particularly clear about the semantics of this Operation, if there are mistakes, thank you for correcting.
> 
> ```
> // RUN: mlir-hlo-opt %s -cse -allow-unregistered-dialect | FileCheck %s
> 
> func.func @rng_bit_generator(%arg0: tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>) {
>   %0 = mhlo.constant dense<[10, 12]> : tensor<2xui64>
>   %1, %2 = "mhlo.rng_bit_generator"(%0) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
>   %3, %4 = "mhlo.rng_bit_generator"(%0) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
>   func.return %1, %2, %3, %4 : tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>
> }
> ```
We will get:
```
module {
  func @rng_bit_generator(%arg0: tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>) {
    %0 = mhlo.constant dense<[10, 12]> : tensor<2xui64>
    %output_state, %output = "mhlo.rng_bit_generator"(%0) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
    return %output_state, %output, %output_state, %output : tensor<2xui64>, tensor<10x12xui32>, tensor<2xui64>, tensor<10x12xui32>
  }
}
```

**Comment by jpienaar** [comment]
We don't expect it to be different with same input state no:

"The output is guaranteed to be a deterministic function of the initial state but it is not guaranteed to be deterministic between backends and different compiler versions."

https://www.tensorflow.org/xla/operation_semantics#rngbitgenerator

If you wanted different numbers you have to use returned updated state from output of one as input to the other.

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops_base_enums.td [hunk 1]
```diff
@@ -173,4 +173,20 @@ def HLO_FusionKind : I32EnumAttr<"FusionKind", "fusion kind", [
...
+    "XLA PRNG algorithm to be used.",
+    [
+      HLO_RNG_ALGORITHM_DEFAULT,
+      HLO_RNG_ALGORITHM_THREE_FRY,
+      HLO_RNG_ALGORITHM_PHILOX
+    ]> {
+  let genSpecializedAttr = 0;
+  let cppNamespace = "::mlir::mhlo";
+}
+def HLO_RngAlgorithmAttr : EnumAttr<HLO_Dialect, HLO_RNG_ALGORITHM, "algorithm">;
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops_base_enums.td:190]: Perhaps "algorithm" -> "rng" ?

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir [hunk 2]
```diff
@@ -965,7 +965,7 @@ func.func @rng_bit_generator(%arg0: tensor<2xui64>) -> (tensor<2xui64>, tensor<1
   %4 = mhlo.constant dense<[10, 12]> : tensor<2xui64>
   %0 = mhlo.constant dense<[10, 12]> : tensor<2xi32>
   %1 = mhlo.constant dense<3> : tensor<i32>
-  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = 0 : i32} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
+  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = #mhlo<"algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir:968]: I think` #mhlo.rng<DEFAULT>` should work here then

**Review by jpienaar — APPROVED** [review]
Nice, we may have to change the Python bindings too, if you are able to that would be great (JAX uses the Python bindings and I don't know off the top of my head if this will change the python build signature).

@hawkinsp for visibility.

**Comment by hawkinsp** [comment]
> Nice, we may have to change the Python bindings too, if you are able to that would be great (JAX uses the Python bindings and I don't know off the top of my head if this will change the python build signature).
> 
> @hawkinsp for visibility.

Yes, I suspect this code will need an update:
https://github.com/google/jax/blob/39ac9faf61e98169d999d33418ca58cecaccc843/jax/_src/lax/lax.py#L4119

unless the bindings still accept i32 attrs in that place for now.

**Comment by hawkinsp** [comment]
> Yes, I will try to modify pybind.Thanks for your remind.

Note: you don't necessarily *have* to maintain backward compatibility for JAX specifically. There is a way we manage breaking changes to MHLO at the moment wrt to JAX. It requires two PRs:
a) a TF PR:
* make your breaking MHLO change in the TF tree.
* increment this version number in the same PR https://github.com/tensorflow/tensorflow/blob/1249b029f84c7424eb059710e6c1fb1dce713d31/tensorflow/compiler/xla/python/xla_client.py#L49

b) a JAX PR, in which you change the code that will be broken to look like this:
```
if jax._src.lib.mlir_api_version >= 1234:
  # new code
else
  # old code
```

and then we submit (b) before (a).

This process could probably be improved, but it does work. Also, it's not necessarily your obligation to do this unless you want to: we can also do the work as part of the TF PR merge process (one thing we do when merging a TF PR is running a number of downstream tests, which would catch this kind of issue and the person doing the merge would make the fix.)

**Comment by jpienaar** [comment]
+1 to what Peter said (nice to have but not blocker)

**Comment by lipracer** [comment]
I see.  I will submit a PR for Jax to compatibily current mhlo api version. then modify and submit this PR. Thanks.

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir (hunk 2 above)
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir]: When I replace  #mhlo<"algorithm DEFAULT"> with #mhlo.rng<DEFAULT> It seem didn't work well.

**Comment by lipracer** [comment]
I have submitted the [PR](https://github.com/google/jax/pull/10477) in Jax Repo.

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo-c/Attributes.h [hunk 3]
```diff
@@ -275,6 +275,22 @@ MLIR_CAPI_EXPORTED bool mlirMhloAttributeIsAFusionKindAttr(MlirAttribute attr);
...
+// parameter.
+MLIR_CAPI_EXPORTED MlirAttribute
+mlirMhloRngAlgorithmAttrGet(MlirContext ctx, const std::string &kind);
+
+// Returns true if the given attribute is a RngAlgorithm attribute.
+MLIR_CAPI_EXPORTED bool
+mlirMhloAttributeIsARngAlgorithmAttr(MlirAttribute attr);
+
+// Returns the rng-algorithm string associated with RngAlgorithm attribute.
+MLIR_CAPI_EXPORTED std::string
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo-c/Attributes.h:291]: We shouldn't be using std::string in the C API, but there is existing behavior like this in this file which is already tracked to be removed and this wouldn't add much cost.

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td [hunk 4]
```diff
@@ -2307,9 +2307,8 @@ def HLO_RngBitGeneratorOp : HLO_Op<"rng_bit_generator", [NoSideEffect]> {
     https://www.tensorflow.org/xla/operation_semantics#rngbitgenerator.
   }];
   let arguments = (ins
-    // TODO(jpienaar): This could be an enum instead.
-    I32Attr:$rng_algorithm,
-    HLO_IntOrFpTensor:$initial_state
+    HLO_IntOrFpTensor:$initial_state,
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:2310]: Why the order change? Prefer to keep this consistent with the argument order in operation semantics linked above.

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir [hunk 5]
```diff
@@ -965,7 +965,7 @@ func.func @rng_bit_generator(%arg0: tensor<2xui64>) -> (tensor<2xui64>, tensor<1
   %4 = mhlo.constant dense<[10, 12]> : tensor<2xui64>
   %0 = mhlo.constant dense<[10, 12]> : tensor<2xi32>
   %1 = mhlo.constant dense<3> : tensor<i32>
-  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = 0 : i32} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
+  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = #mhlo<"rng_algorithm DEFAULT">} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
   func.return %2, %3 : tensor<2xui64>, tensor<10x12xui32>
```
Comment (jpienaar) [inline tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir:969]: What error do you get with `mhlo.rng_algorithm<DEFAULT>`? (No quotes)

Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir:969]: error message:
```
error: unexpected error: expected string literal data in dialect symbol
  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = #mhlo<DEFAULT>} : (tensor<2xui64>) -> (tensor<2xui64>, tensor<10x12xui32>)
```
and this is quotes version:
```
error: unexpected error: unknown mhlo attribute
  %2, %3 = "mhlo.rng_bit_generator"(%4) {rng_algorithm = #mhlo<"DEFAULT">} : (tensor<2xui64>) -> (tensor<3xui64>, tensor<10x12xui32>)
```

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td (hunk 4 above)
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:2310]: I have the following reasons:
1. attibuts always declare follow by operand.
2. xla hlo builder declare algorithm argument after state.

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo-c/Attributes.h (hunk 3 above)
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo-c/Attributes.h:291]: I will change to const char*.

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td (hunk 4 above)
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:2310]: I refer to the statement [here](https://github.com/tensorflow/tensorflow/blob/19025b6feacae8ef6636e0166e21ae5c2f04be64/tensorflow/compiler/xla/service/hlo_instruction.h#L566).I'm going to change those back.

Comment (jpienaar) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td:2310]: Yeah its one of those unfortunate things: the one you pointed out is "server side HLO" and considered internal. But it is confusing indeed.

File: tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir (hunk 5 above)
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/tests/Dialect/mhlo/ops.mlir:969]: I refer to the definition of MLIR Core.Maybe we need assemblyFormat.
```
def HLO_RngAlgorithmAttr : EnumAttr<HLO_Dialect, HLO_RNG_ALGORITHM, "rng_algorithm"> {
  let assemblyFormat = "`<` $value `>`";
}
```
then use ```{rng_algorithm = #mhlo.rng_algorithm<DEFAULT>}``` it work well.

**Review by jpienaar — APPROVED** [review]
Much nicer, thanks!

**Comment by jpienaar** [comment]
Can you check tests:

//tensorflow/compiler/mlir/xla/tests:legalize-tf.mlir.test
//tensorflow/compiler/mlir/xla/tests/translate:export.mlir.test
//tensorflow/compiler/mlir/xla/tests/translate:import.hlotxt.test

?

**Comment by lipracer** [comment]
Yes. I just run with command:
```bazelisk test tensorflow/compiler/mlir:all --cache_test_results=no```
and all pass, It seem some test no run.

**Review by jpienaar — APPROVED** [review]

**Comment by lipracer** [comment]
I have resubmitted jax's [patch](https://github.com/google/jax/pull/10482), but it has not been merged yet.Integration testing may also be affected.

**Comment by jpienaar** [comment]
SG, I didn't see any of those failing here (I mentioned the other set yesterday and those you seem to have fixed but kokoro pending).

**Review by hawkinsp — APPROVED** [review]
The JAX PR was merged, so I think this is ready to merge.

**Comment by lipracer** [comment]
Thanks.

**Comment by lipracer** [comment]
Some parameters still refer to std::string, I'm going to clean them all up in another patch.

**Comment by hawkinsp** [comment]
I think you actually need to fix the `std::string` references before merging: they cannot appear in a C API.

**Comment by lipracer** [comment]
cleanup std::string in attributes file other [PR](https://github.com/tensorflow/tensorflow/pull/55888)

**Review by hawkinsp — APPROVED** [review]

File: tensorflow/compiler/mlir/hlo/lib/CAPI/Attributes.cpp [hunk 6]
```diff
@@ -510,3 +510,27 @@ std::string mlirMhloFusionKindAttrGetFusionKind(MlirAttribute attr) {
...
+  return wrap(
+      mlir::mhlo::RngAlgorithmAttr::get(unwrap(ctx), rng_algorithm.getValue()));
+}
+
+bool mlirMhloAttributeIsARngAlgorithmAttr(MlirAttribute attr) {
+  return unwrap(attr).isa<mlir::mhlo::RngAlgorithmAttr>();
+}
+
+const char *mlirMhloRngAlgorithmAttrGetRngAlgorithm(MlirAttribute attr) {
+  return mlir::mhlo::stringifyRngAlgorithm(
```
[the remark this answers is not in the document]
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/lib/CAPI/Attributes.cpp:532]: Thanks, ```StringRef::str()``` will create temporary std::string indeed.

**Review by hawkinsp — APPROVED** [review]

**Review by hawkinsp — APPROVED** [review]

**Pull request merged**